src/Moq/SetupCollection.cs

```diff
@@ -11,17 +11,35 @@ internal sealed class SetupCollection
 	{
 		private List<Setup> setups;
 		private uint overridden;  // bit mask for the first 32 setups flagging those known to be overridden
+		private bool hasEventSetup;
```

Comment:
Something about this flag is bothering me&mdash;perhaps its being an overly specific flag&mdash;and I would prefer if we didn't have to add this to `SetupCollection`.

If I understand its purpose correctly (see also my other review comment), it's some kind of backward-compatibility flag to determine whether event handler subscriptions & unsubscriptions should be recorded for verification. If so, it might be more suitable to have this flag sit on the `Mock` instance, more specifically in `Mock.Switches`: `mock.Switches |= Switches.RecordEventAccessorInvocations`. (TBH I don't fancy that solution a lot either, because I think I made an API design mistake by introducing `Mock.Switches`. Perhaps we'd want to hide that switch from public view using `[EditorBrowsable]` in order to not encourage `Switch` much further... but I'm not sure about this.)

Do you have any thoughts on this?

Reply:
TBH I didn't know about `Switches`. Looking at them it seems they are designed to enable\disable a mock feature while initializing. Are they appropriate for such a scenario where we tweak something during the lifetime of the mock? Few things about `Switches`:
- Currently, it has a public setter, potentially this could be dangerous. Imagine someone starts playing with this flag externally while you are in the middle of handling of the invocation. This could, for example, lead to prematurely returned invocation. (I do not think this is a realistic scenario though, probably nobody will do that)
- The logic will be spread a bit. It means that whenever SetupCollection has been touched the flag could become invalid. In other words, whenever SetupCollection is changed the flag should be revisited as well.
- There is one advantage. That we could prevent the following: Currently, because of the lack of F# support, we could not use `IsSpecialMethod` while checking if a method is an event. This could lead to a bit awkward situation where a method could be recognized as an event even though it isn't. If, for example, someone name method `add_Someting(int a)` or `remove_Something(int b)` the `hasEventSetup` will be set to `true`. This cannot make any functional problems but could slightly impact the performances while handling event invocation. However, I don't think this is a real threat as I cannot imagine there is anyone in this world who will name method starting with `add_` or `remove_` 👿 

I do not prefer any of the solutions. Seems it fits well into `SetupCollection` but on the other hand it is a backward compatibility flag and we are polluting `SetupCollection` a bit.

Reply:
Sorry for the late feedback on this one. I don't have a clear favourite solution at the moment, so I suggest that we go along with your solution, i.e. a `hasEventSetup` flag inside `SetupCollection`, as it is good enough for what we want to achieve right now. We can still a refactor this in the future.